PlusRemote/MRML/vtkMRMLPlusRemoteNode.cxx

```diff
@@ -0,0 +1,616 @@
+  {
+    roiOrigin[i] = roiCenter[i] - roiRadius[i];
+  }
+
+  double outputSpacing = self->GetLiveReconstructionSpacing();
+  int outputDimensions[3] = {
+    std::ceil((2 * roiRadius[0]) / outputSpacing),
+    std::ceil((2 * roiRadius[1]) / outputSpacing),
+    std::ceil((2 * roiRadius[2]) / outputSpacing)
+  };
+
+  self->SetLiveReconstructionROIDimensions(outputDimensions);
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLPlusRemoteNode::SetLiveReconstructionDisplayROI(bool displayROI)
+{
+  vtkMRMLAnnotationROINode* roiNode = this->GetLiveReconstructionROINode();
+  if (roiNode)
+  {
+    roiNode->SetDisplayVisibility(displayROI);
+  }
+
+  bool oldVisibility = this->LiveReconstructionDisplayROI;
+  this->LiveReconstructionDisplayROI = displayROI;
+  if (oldVisibility != displayROI)
+  {
+    this->Modified();
+  }
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLPlusRemoteNode::SetLiveReconstructionSpacing(double spacing)
+{
+  double oldSpacing = this->LiveReconstructionSpacing;
+  if (oldSpacing == spacing)
+  {
+    return;
+  }
+  int wasModifying = this->StartModify();
+
+  this->LiveReconstructionSpacing = spacing;
+
+  int oldDimensions[3] = { 0, 0, 0 };
+  this->GetLiveReconstructionROIDimensions(oldDimensions);
+
+  int newDimensions[3] = { 0, 0, 0 };
+  for (int i = 0; i < 3; ++i)
+  {
+    newDimensions[i] = std::ceil((oldSpacing*oldDimensions[i]) / spacing);
+  }
+  this->SetLiveReconstructionROIDimensions(newDimensions);
+  this->Modified();
+  this->EndModify(wasModifying);
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLPlusRemoteNode::SetLiveReconstructionROIDimensions(int x, int y, int z)
+{
+  int dimensions[3] = { x, y, z };
+  this->SetLiveReconstructionROIDimensions(dimensions);
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLPlusRemoteNode::SetLiveReconstructionROIDimensions(int dimensions[3])
+{
+  bool modified = false;
+  for (int i = 0; i < 3; ++i)
+  {
+    if (dimensions[i] != this->LiveReconstructionROIDimensions[i])
+    {
+      modified = true;
+      break;
+    }
+  }
+
+  if (!modified)
+  {
+    return;
+  }
+
+  int wasModifying = this->StartModify();
+
+  for (int i = 0; i < 3; ++i)
+  {
+    this->LiveReconstructionROIDimensions[i] = dimensions[i];
+  }
+
+  vtkMRMLAnnotationROINode* roiNode = this->GetLiveReconstructionROINode();
+  if (roiNode)
+  {
+    double spacing = this->LiveReconstructionSpacing;
+    double radius[3] = { 0.0, 0.0, 0.0 };
+    for (int i = 0; i < 3; ++i)
+    {
+      radius[i] = (dimensions[i] * spacing) / 2;
+    }
+    roiNode->SetRadiusXYZ(radius);
+  }
+  this->Modified();
+  this->EndModify(wasModifying);
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLPlusRemoteNode::WriteXML(ostream& of, int nIndent)
```

Comment:
Use new macros for read/write XML, print, and copy attributes:
vtkMRMLWriteXML…
vtkMRMLReadXML…
vtkMRMLCopy…
vtkMRMLPrint...

Reply:
I used the macros where possible, but I couldn't adapt some attributes that are stored in std::vector.